contracts/src/ZKube.sol

```diff
@@ -72,19 +73,43 @@ contract ZKube is IZKube {
         game = games[id];
         roundBlockNumber = getBlock(game.interval, game.startingBlock, game.numberOfRounds);
         uint256 randomNumber = getRandomNumber(roundBlockNumber);
-        puzzle = IZKubePuzzleSet(game.puzzleSet).getPuzzle(randomNumber);
+        puzzle = IZKubePuzzleSet(game.puzzleSet).getRandomPuzzle(randomNumber);
+    }
+
+    function parseInputSignals(Puzzle memory puzzle, address sender) public pure returns(uint256[137] memory inputSignals) {
+        uint[] memory base4StartingGrid = puzzle.startingGrid.hexToBase4();
+        uint[] memory base4FinalGrid = puzzle.finalGrid.hexToBase4();
+
+        for (uint256 i = 0; i < 64; i++) {
+            inputSignals[i] = uint8(base4StartingGrid[i]);
+            inputSignals[i+64] = uint8(base4FinalGrid[i]);
+        }
+        for (uint256 j = 0; j < 8; j++) {
+            inputSignals[j+128] = puzzle.availableFunctions[j];
+        }
+        inputSignals[136] = uint160(sender);
```

Comment:
minor optimisation to call msg.sender here directly rather than passing around the sender address as function param